### PR TITLE
[Feature] Add ASTER

### PR DESCRIPTION
**Pull request by GHuiXin**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation
There is a recent paper 《ASTER: An Attentional Scene Text Recognizer with Flexible Rectification》, which is a very classic paper. its ideas are useful guidelines for character recognition, so to do code reproduction.

## Modification

Done:
- Aster_encoder
- Aster_decoder
- Aster_decoder_with_beamsearch
- Attention_based LSTM
- README.md
- metafile.yml(Part)

## BC-breaking (Optional)
  In origin mmocr,  I found two questions , as follows:
- the strides of kernel is irregular, so the assert(assert utils.is_type_list(strides, int)) should be commented in resnet_abi.py.
-  # decoder.update(num_classes=self.label_convertor.num_classes())
   # decoder.update(start_idx=self.label_convertor.start_idx)
   # decoder.update(padding_idx=self.label_convertor.padding_idx)
   They were annotated by me
 but i didn't modified origin codes in PR. 
## Checklist

**Before PR**:

- [x] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmocr/blob/main/.github/CONTRIBUTING.md) to create this PR.
- [x] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmocr/blob/main/.github/CONTRIBUTING.md) are used to fix the potential lint issues.
- [ ] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with some of those projects.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

**Comment by gaotongxiao** [comment]
Hi, please follow the guide to lint your code: https://github.com/open-mmlab/mmocr/blob/main/.github/CONTRIBUTING.md#installing-pre-commit-hooks

**Comment by GHuiXin** [comment]
> Hi, please follow the guide to lint your code: https://github.com/open-mmlab/mmocr/blob/main/.github/CONTRIBUTING.md#installing-pre-commit-hooks
OK

**Comment by xinke-wang** [comment]
Hi @GHuiXin, Thank you for your contribution! We'll review this PR soon!

File: configs/_base_/recog_models/asternet.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
+# model
+label_convertor = dict(
+    type='AttnConvertor', dict_type='DICT36', with_unknown=True, lower=True)
+
+model = dict(
+    type='ASTERNet',
```
Comment (Mountchicken) [inline configs/_base_/recog_models/asternet.py:6]: We can just call it ASTER

File: configs/_base_/recog_models/asternet.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    preprocessor=dict(
+        type='TPSPreprocessor',
+        img_size=(32, 100),
+        rectified_img_size=(32, 100),
+        num_img_channel=3),
+    backbone=dict(
+        type='ResNetABI',
+        arch_settings=[3, 4, 6, 6, 3],
+        strides=[2, 2, [2, 1], [2, 1], [2, 1]],
+        init_cfg=dict(type='Xavier', layer='Conv2d')),
```
Comment (Mountchicken) [inline configs/_base_/recog_models/asternet.py:16]: We have recently refactored the [ResNet structure](https://github.com/open-mmlab/mmocr/pull/809). ResNet in ASTER can be implemented using this configuration
```python
backbone = dict(
        type='ResNet'
        in_channels=3,
        stem_channels=[64, 128],
        block_cfgs=dict(type='BasicBlock', use_conv1x1='True'),
        arch_layers=[3, 4, 6, 6, 3],
        arch_channels=[32, 64, 128, 256, 512],
        strides=[(2, 2), (2, 2), (2, 1), (2, 1), (2, 1)])
```

File: mmocr/models/textrecog/decoders/aster_decoder.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+import torch
+import torch.nn.functional as F
+
+from mmocr.models.builder import DECODERS
+from mmocr.models.textrecog.layers import attention_aster_layer
+from .base_decoder import BaseDecoder
+
+
+@DECODERS.register_module()
+class ASTERDecoder(BaseDecoder):
```
Comment (Mountchicken) [inline mmocr/models/textrecog/decoders/aster_decoder.py:11]: - ASTER uses the Attention decoder, and I believe there should be a corresponding implementation in mmocr. 
- @gaotongxiao [SequenceAttentionDecoder](https://github.com/open-mmlab/mmocr/blob/52c95ccaac/mmocr/models/textrecog/decoders/sequence_attention_decoder.py) seems to be a generic Attention decoder? MMOCR has so many Attention-based algorithms, there should be a generic Attention decoder, just like our recently reviewed Master, there is a generic transformer decoder.

File: mmocr/models/textrecog/layers/attention_aster_layer.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+
+class AttentionLSTM(nn.Module):
```
Comment (Mountchicken) [inline mmocr/models/textrecog/layers/attention_aster_layer.py:7]: The same as aster_decoder, there is already a corresponding implementation in MMOCR

File: mmocr/models/textrecog/encoders/aster_encoder.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+from mmcv.runner import Sequential
+
+from mmocr.models.builder import ENCODERS
+from mmocr.models.textrecog.layers import BidirectionalLSTM
+from .base_encoder import BaseEncoder
+
+
+@ENCODERS.register_module()
+class ASTEREncoder(BaseEncoder):
```
Comment (Mountchicken) [inline mmocr/models/textrecog/encoders/aster_encoder.py:10]: @gaotongxiao  We can refactor this to LSTMEncoder. BiLSTM is not only used in ASTER, but also in other algorithms such as CRNN.

**Pull request closed without merging**